crates/typed-store/src/metrics.rs

```diff
@@ -1,12 +1,103 @@
 // Copyright (c) 2022, Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
-use std::sync::Arc;
-
 use once_cell::sync::OnceCell;
 use prometheus::{
     exponential_buckets, register_histogram_vec_with_registry,
     register_int_counter_vec_with_registry, HistogramVec, IntCounterVec, Registry,
 };
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::time::{Duration, UNIX_EPOCH};
+use std::{sync::Arc, time::SystemTime};
+
+#[derive(Debug, Default, Clone)]
+/// A struct for keeping track of the sampling
+pub struct PerfSampler {
+    // Number of operations since last sample
+    pub num_ops: Arc<AtomicU64>,
+    // Timestamp of the last sample
+    pub latest_sample_ts: Arc<AtomicU64>,
+}
+
+#[derive(Debug, Clone)]
+// A struct for sampling based on number of operations or duration.
+// Sampling happens if the duration expires or number of operations
+// is more than the threshold
+pub struct SamplingInterval {
+    // Sample once every time duration
+    pub once_every_duration: Duration,
+    // Sample once every number of operations
+    pub after_num_ops: u64,
+    // Perf sampler for keeping track of previous sample
+    pub perf_sampler: PerfSampler,
+}
+
+impl Default for SamplingInterval {
+    fn default() -> Self {
+        SamplingInterval {
+            once_every_duration: Duration::from_secs(1),
+            after_num_ops: u64::MAX,
+            perf_sampler: PerfSampler {
+                num_ops: Arc::new(AtomicU64::new(0)),
+                latest_sample_ts: Arc::new(AtomicU64::new(0)),
+            },
+        }
+    }
+}
+
+impl SamplingInterval {
+    pub fn new(once_every_duration: Duration, after_num_ops: u64) -> Self {
+        SamplingInterval {
+            once_every_duration,
+            after_num_ops,
+            perf_sampler: PerfSampler {
+                num_ops: Arc::new(AtomicU64::new(0)),
+                latest_sample_ts: Arc::new(AtomicU64::new(0)),
+            },
+        }
+    }
+    pub fn sample(&self) -> bool {
+        if self.after_num_ops == u64::MAX && self.once_every_duration == Duration::MAX {
+            return false;
+        }
+        if self.after_num_ops != u64::MAX {
+            let num_ops = self.perf_sampler.num_ops.fetch_add(1, Ordering::Relaxed);
+            if num_ops >= self.after_num_ops {
+                if num_ops > self.after_num_ops {
+                    // Multiple threads got here at the same time, only one
+                    // should be able to get the sample
+                    return false;
+                }
+                let now = SystemTime::now()
```

Comment:
I've been told that `Instant::now()` is faster and much preferable for measuring elapsed time compared to SystemTime.

Reply:
The PR further down the https://github.com/MystenLabs/mysten-infra/pull/181 actually uses coarsetime::Instant which is faster and less accurate and doesn't involve a system call. I am going to bring it up here to make it easy to review